Added examples for 3D space charge benchmarking.

cemitch99 opened the pull request:
- [x] depends on #162
- [x] adjust resolution via `amr.n_cell` and `geometry.prob_relative`
- [x] add Python equivalent inputs

The following examples were added to test 3D space charge:

1) A constant focusing channel with an intense waterbag (6D) beam - beam is rms-matched and second moments should be well-preserved.
2) A constant focusing channel with an intense Kurth (6D) beam - beam is rms-matched and distribution should be fully stationary.
3) A periodic focusing channel with an intense Kurth (6D) beam - beam is rms-matched and distribution should vary with the periodicity of the lattice.

Corresponding to 3), an example was added for the same lattice, but without space charge:

4) A periodic focusing channel with an zero-charge Kurth (6D) beam - beam is rms-matched and distribution should vary with the periodicity of the lattice.

Follow-ups commits:

- [x] Add example documentation (partially complete).
- [ ] Add example using a FODO lattice + short rf cavities.

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -187,3 +187,17 @@ add_test(NAME iotalens.analysis
+# Kurth Beam in a Periodic Channel Test ###################################################
+#
+add_test(NAME kurth_periodic.run
+         COMMAND $<TARGET_FILE:app>${ImpactX_SOURCE_DIR}/examples/kurth/input_kurth_periodic.in
```
Little typo that fails the test:
```suggestion
         COMMAND $<TARGET_FILE:app> ${ImpactX_SOURCE_DIR}/examples/kurth/input_kurth_periodic.in
```

ax3l commented:
Thanks a lot, @cemitch99 !

Do we need to increase tolerances for these tests? I think they should fail as long as space charge is not enabled, right?

cemitch99 commented:
> Thanks a lot, @cemitch99 !
> 
> Do we need to increase tolerances for these tests? I think they should fail as long as space charge is not enabled, right?

Yes, they should fail if space charge is not enabled.  I will modify the tolerances now.

ax3l commented:
Thanks. Looks like we need to be even stricter :)

ax3l commented:
Hi @cemitch99, can you please check if the updates make sense?

I think there is a 4. th test (another Kurth addition) that is not mentioned in the PR description yet. Can you also check all 4 Kurth tests are also described in the README.rst of `Kurth`? I think that section is not too clear to me yet.

ax3l commented on examples/kurth/input_kurth_10nC.in:
```diff
@@ -0,0 +1,37 @@
+beam.charge = 1.0e-8
+beam.particle = proton
+beam.distribution = kurth6d
+beam.sigmaX = 1.2154443728379865788e-3
```
double check vs. docs: Are LHS of `amrex::ParmParse` (keys) case-sensitive?
(Since we document and query all sigmas as lower case.)

ax3l replied:
The logic for those params should not be case insensitive, but we use this already in:
- `fodo_rf`
- `multipole`
- `chicane`
- `kurth`
- ...
:thinking:

ax3l replied:
Oh, I think it's a documentation mistake in ImpactX

ax3l replied:
-> #258

cemitch99 committed:
Update analysis_cfchannel_10nC.py

Modified tolerance slightly

cemitch99 committed:
Update input_cfchannel_10nC.in

Selected amr.n_cell, lattice.nslice parameters, and geometry.prob_relative parameters.

cemitch99 committed:
Update input_kurth_periodic.in

Added nonzero charge.  However, this is not a space charge test.  It is, rather, another distribution/lattice test.  It should be run algo.space_charge = false.

cemitch99 committed:
Update analysis_kurth_periodic.py

Modified tolerance slightly.

cemitch99 committed:
Update input_kurth_10nC_periodic.in

Added appropriate numerical parameters for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Update analysis_kurth_10nC_periodic.py

1) Improved the accuracy of the target values used for analysis, 2) modified the tolerance slightly

cemitch99 committed:
Update analysis_kurth_periodic.py

Improved the accuracy of the target values.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Update analysis_kurth_10nC.py

Modified the tolerance slightly.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_spacecharge_benchmarks

ax3l commented:
Hi @cemitch99,

I prepared the branch so far and we just need to make the tolerances a bit more strict I think.
These tests pass even if space-charge is off
- `cfchannel_spacecharge.analysis`
- `kurth_periodic.analysis` (update: ah, has no space charge)

The other two are good: fail if space charge is off and pass if turned on :)

cemitch99 committed:
Update analysis_kurth_periodic.py

Tightened the tolerance.  This test should pass if space charge is OFF, and fail is space charge is ON.

cemitch99 committed:
Update input_kurth_periodic.in

Increased the default charge to match the value in input_kurth_10nC_periodic.in.  This test should be run with space charge OFF.  If it is run with space charge ON, then it will fail with large error.

ax3l committed:
ConstF Channel: Python 10nC Example

ax3l commented on examples/kurth/README.rst:
```diff
@@ -48,3 +48,54 @@ We run the following script to analyze correctness:
       :caption: You can copy this file from ``examples/kurth/analysis_kurth.py``.
+
+
+.. _examples-kurth-10nC:
```
@cemitch99 I am a bit confused here - we have three new inputs files for tests in this PR but only document one. Do we need to add more info here?

ax3l committed:
Python: kurth periodic example

cemitch99 commented on examples/kurth/input_kurth.in:
```diff
@@ -35,3 +32,6 @@ constf1.kt = 1.0
 algo.space_charge = false
+
+amr.n_cell = 40 40 32
+geometry.prob_relative = 1.0
```
The above test, "input_kurth.in", uses only a Kurth distribution and a constant focusing element (without space charge), features that are already tested elsewhere.  I suggest that this test can be removed to save time.  The two important Kurth tests are "input_kurth_periodic.in" (without space charge) and "input_kurth_10nC_periodic.in" (with space charge).

ax3l replied:
I agree - thanks for the update, I will add one more push that removes `input_kurth.in` :)

cemitch99 commented on examples/kurth/input_kurth_10nC.in:
```diff
@@ -0,0 +1,41 @@
+
+amr.n_cell = 48 48 40
+#amr.n_cell = 72 72 72 #optional for increased precision
+geometry.prob_relative = 1.0
```
The above test, "input_kurth_10nC.in", uses only a Kurth distribution and a constant focusing element (with space charge), features that are already tested elsewhere.  I suggest that this test can be removed to save time.  The two important Kurth tests are "input_kurth_periodic.in" (without space charge) and "input_kurth_10nC_periodic.in" (with space charge).

ax3l replied:
I agree - thanks for the update, I will add one more push that removes `input_kurth_10nC.in` :)

cemitch99 committed:
Update README.rst

Modified documentation of the relevant benchmarks tests using the Kurth distribution, "input_kurth_periodic.in" and "input_kurth_10nC_periodic.in".  (One without space charge, and one with space charge.)

cemitch99 commented on examples/CMakeLists.txt:
```diff
@@ -354,6 +373,41 @@ add_impactx_test(iotalattice.py.MPI
+
+# w/ space charge
+add_impactx_test(kurth_10nC
+    examples/kurth/input_kurth_10nC.in
```
Suggest that the test here, corresponding to "input_kurth_10nC.in", can be removed since the previous test "input_kurth_10nC_periodic.in" is more comprehensive.

ax3l replied:
Thanks - removed :)

ax3l committed:
Remove Already Covered Scenarios

See Chad's GitHub comments for context.

ax3l committed:
Python: kurth 10nC periodic
